kvserver: improve observability with decommission nudger

wenyihu6 opened the pull request:
Stacked on top of https://github.com/cockroachdb/cockroach/pull/152792
Resolves: https://github.com/cockroachdb/cockroach/issues/151847
Epic: none

---

**kvserver: improve observability with decommission nudger**

Previously, we added the decommissioning nudger which nudges the leaseholder
replica of decommissioning ranges to enqueue themselves into the replicate queue
for decommissioning. However, we are still observing extended decommission stall
with the nudger enabled. Observability was limited, and we could not easily tell
whether replicas were successfully enqueued or processed.

This commit improves observability by adding four metrics to track the enqueue
and processing results of the decommissioning nudger:
ranges.decommissioning.nudger.{enqueue,process}.{success,failure}.

---

**kvserver: add enqueue metrics to base queue**

Previously, observability into base queue enqueuing was limited to pending queue
length and process results. This commit adds enqueue-specific metrics for the
replicate queue:

- queue.replicate.enqueue.add: counts replicas successfully added to the queue
- queue.replicate.enqueue.failedprecondition: counts replicas that failed the
  replicaCanBeProcessed precondition check
- queue.replicate.enqueue.noaction: counts replicas skipped because ShouldQueue
  determined no action was needed
- queue.replicate.enqueue.unexpectederror: counts replicas that were expected to
  be enqueued (ShouldQueue returned true or the caller attempted a direct enqueue)
  but failed due to unexpected errors

---

**kvserver: move bq.enqueueAdd update to be outside of defer**

Previously, we updated bq.enqueueAdd inside the defer statement of addInternal.
This was incorrect because we may return queued = true for a replica already
processing and was marked for requeue. That replica would later be requeued in
finishProcessingReplica, incrementing the metric again, lead to double counting.

---

**kvserver: test metrics in TestBaseQueueCallback* and TestReplicateQueueDecommissionScannerDisabled**

his commit extends TestBaseQueueCallback* and
TestReplicateQueueDecommissionScannerDisabled to also verify metric updates.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152787)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: improve observability with decommission nudger

Previously, we added the decommissioning nudger which nudges the leaseholder
replica of decommissioning ranges to enqueue themselves into the replicate queue
for decommissioning. However, we are still observing extended decommission stall
with the nudger enabled. Observability was limited, and we could not easily tell
whether replicas were successfully enqueued or processed.

This commit improves observability by adding four metrics to track the enqueue
and processing results of the decommissioning nudger:
ranges.decommissioning.nudger.{enqueue,process}.{success,failure}.

tbg reviewed (changes requested):
Would like to clarify the metric semantics better, but directionally :thumbsup: 

> @tbg reviewed 1 of 3 files at r2, 4 of 5 files at r3, 2 of 2 files at r4, 3 of 3 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152787#-:-OZ3a7ss975hPAf6htUJ:bpidjtu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/metrics.go` line 2181 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152787#-OZ3ZES_A8JUBucrtTLb:-OZ3ZES_A8JUBucrtTLc:b-9ef2sm) ([raw file](https://github.com/cockroachdb/cockroach/blob/344ecc232d85d283f60d0422615b3f3875340a6b/pkg/kv/kvserver/metrics.go#L2181)):*
> ```Go
> 		Unit:        metric.Unit_COUNT,
> 	}
> 	metaReplicateQueueEnqueueFailures = metric.Metadata{
> ```

what does it mean to fail to be enqueued? In particular, where does the case where `shouldQueue` returns false land? That shouldn't really be called a "failure"
This avoids the distinction between `failed` and `skipped` which I haven't been able to understand even after reviewing the code.

In my understanding, we are interested in distinguishing three cases:

- the replica was added to the queue.
- it was not added to the queue because there was nothing to do for this replica (shouldQueue returned false)
- it was not added to the queue because one of the many preconditions (zone config available, holds lease, etc, did not hold).

So maybe

- `*.enqueue.accepted`
- `*.enqueue.no_action`
- `*.enqueue.failed_precondition`

I'm not totally in love with these names either, but it might be clearer.
___
*[`pkg/kv/kvserver/metrics.go` line 2190 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152787#-OZ3YS7l1-H3vpph4fKW:-OZ3YS7l1-H3vpph4fKX:b-lbp0hm) ([raw file](https://github.com/cockroachdb/cockroach/blob/344ecc232d85d283f60d0422615b3f3875340a6b/pkg/kv/kvserver/metrics.go#L2190)):*
> ```Go
> 	metaReplicateQueueEnqueueSkipped = metric.Metadata{
> 		Name: "queue.replicate.enqueue.skipped",
> 		Help: "Number of replicas which didn't attempt to be enqueued but returned " +
> ```

what does it mean to "not attempt to be enqueued" but then to be skipped?
___
*[`pkg/kv/kvserver/queue.go` line 829 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152787#-OZ3YjJV10ts0juD8hCZ:-OZ3YjJV10ts0juD8hC_:b-i6w01q) ([raw file](https://github.com/cockroachdb/cockroach/blob/344ecc232d85d283f60d0422615b3f3875340a6b/pkg/kv/kvserver/queue.go#L829)):*
> ```Go
> ) (queued bool, err error) {
> 	defer func() {
> 		bq.updateMetricsOnEnqueueResult(queued)
> ```

Could you add a note that `queued => err == nil`, which seems to be true based on reading the method.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: add enqueue metrics to base queue

Previously, observability into base queue enqueuing was limited to pending queue
length and process results. This commit adds enqueue-specific metrics for the
replicate queue:

- queue.replicate.enqueue.add: counts replicas successfully added to the queue
- queue.replicate.enqueue.failedprecondition: counts replicas that failed the
  replicaCanBeProcessed precondition check
- queue.replicate.enqueue.noaction: counts replicas skipped because ShouldQueue
  determined no action was needed
- queue.replicate.enqueue.unexpectederror: counts replicas that were expected to
  be enqueued (ShouldQueue returned true or the caller attempted a direct enqueue)
  but failed due to unexpected errors

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152787)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/metrics.go` line 2181 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152787#-OZ3ZES_A8JUBucrtTLb:-OZBxrWn5HLq7XxRc_7G:blx9sgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/344ecc232d85d283f60d0422615b3f3875340a6b/pkg/kv/kvserver/metrics.go#L2181)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

what does it mean to fail to be enqueued? In particular, where does the case where `shouldQueue` returns false land? That shouldn't really be called a "failure"
This avoids the distinction between `failed` and `skipped` which I haven't been able to understand even after reviewing the code.

In my understanding, we are interested in distinguishing three cases:

- the replica was added to the queue.
- it was not added to the queue because there was nothing to do for this replica (shouldQueue returned false)
- it was not added to the queue because one of the many preconditions (zone config available, holds lease, etc, did not hold).

So maybe

- `*.enqueue.accepted`
- `*.enqueue.no_action`
- `*.enqueue.failed_precondition`

I'm not totally in love with these names either, but it might be clearer.
</blockquote></details>

Good points, semantics were unclear. I’ve reverted the commit and added four metrics with definitions to clarify their semantics. Lmk if this aligns with what you had in mind (the only new case here is unexpected error)

```
- queue.replicate.enqueue.add: counts replicas successfully added to the queue  
- queue.replicate.enqueue.failedprecondition: counts replicas that failed the  
  replicaCanBeProcessed precondition check  
- queue.replicate.enqueue.noaction: counts replicas skipped because ShouldQueue  
  determined no action was needed  
- queue.replicate.enqueue.unexpectederror: counts replicas that were expected to  
  be enqueued (ShouldQueue returned true or the caller attempted a direct enqueue)  
  but failed due to unexpected errors
```
___
*[`pkg/kv/kvserver/metrics.go` line 2190 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152787#-OZ3YS7l1-H3vpph4fKW:-OZByQRE3jK_lVtn3r_v:b-y0pgbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/344ecc232d85d283f60d0422615b3f3875340a6b/pkg/kv/kvserver/metrics.go#L2190)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

what does it mean to "not attempt to be enqueued" but then to be skipped?
</blockquote></details>

Agree that the wording was confusing. Hope it's clearer now. 
___
*[`pkg/kv/kvserver/queue.go` line 829 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152787#-OZ3YjJV10ts0juD8hCZ:-OZByq0v5LgG01AunaJ9:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/344ecc232d85d283f60d0422615b3f3875340a6b/pkg/kv/kvserver/queue.go#L829)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Could you add a note that `queued => err == nil`, which seems to be true based on reading the method.
</blockquote></details>

Added.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> @tbg reviewed 4 of 4 files at r7, 3 of 3 files at r8, 4 of 4 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152787)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/queue_test.go` line 1301 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152787#-OZDm6aZ1XfxrrksTorV:-OZDm6aZ1XfxrrksTorW:b-doyub6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0eaceacf8b1cc1ebf47300b17fb2e947183aede/pkg/kv/kvserver/queue_test.go#L1301)):*
> ```Go
> // 7. processing: the replica is already being processed and not enqueued again.
> // 8. full queue: the queue is full and the replica is not enqueued again.
> func TestBaseQueueCallbackOnEnqueueResult(t *testing.T) {
> ```

Can the new metrics be sanity-checked in this test as well?


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: move bq.enqueueAdd update to be outside of defer

Previously, we updated bq.enqueueAdd inside the defer statement of addInternal.
This was incorrect because we may return queued = true for a replica already
processing and was marked for requeue. That replica would later be requeued in
finishProcessingReplica, incrementing the metric again, lead to double counting.

wenyihu6 committed:
kvserver: test metrics in TestBaseQueueCallback* and TestReplicateQueueDecommissionScannerDisabled

his commit extends TestBaseQueueCallback* and
TestReplicateQueueDecommissionScannerDisabled to also verify metric updates.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152787)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/queue_test.go` line 1301 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152787#-OZDm6aZ1XfxrrksTorV:-OZFn25060Ds6XnssOGU:b-i68uuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0eaceacf8b1cc1ebf47300b17fb2e947183aede/pkg/kv/kvserver/queue_test.go#L1301)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Can the new metrics be sanity-checked in this test as well?
</blockquote></details>

I found a double counting flaw in my metrics updates while adding the tests - we might return queued = true from addInternal for a processing replica that was marked as requeued. I pushed a fix for it as a follow up if you wanna have a look. 


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> @tbg reviewed 1 of 1 files at r11, 2 of 2 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152787)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Rebased on master to pick up https://github.com/cockroachdb/cockroach/pull/152967 which caused flakes on CI.

wenyihu6 commented:
TFTR!

bors r=tbg
